plugins/s3_auth/s3_auth.cc

```diff
@@ -391,6 +451,15 @@ class S3Config
     TSHttpTxnHookAdd(txnp, TS_HTTP_SEND_REQUEST_HDR_HOOK, _cont);
   }
 
+  void
+  schedule_conf_reload(long delay) const
+  {
+    TSContScheduleOnPool(_conf_rld, delay * 1000, TS_THREAD_POOL_NET);
+  }
+
+  std::shared_mutex reload_mutex;
```

Comment:
This locking strategy looks like it could block multiple event threads while a new config is being loaded.  I suggest letting transactions use the old config while the new one is being loaded, using https://godbolt.org/z/3bq43z or something like it.

Reply:
I like the idea, but to use this it seems that the whole `config` class needs to be re-written to work with it, feels like a clean up PR rather than a simple feature one.

Reply:
I don't see the need for a major rewrite of the config class to use PingPong.  But the version of PingPong above initializes the instance of T to be written from the last one, which is not good in all cases, and not good with the config class.  PingPong can be changed like this:  https://godbolt.org/z/c4qhfW
I think with this version of PingPong, the only change to the config class (to use it for the T parameter) would be to move the two continuations outside of the class instance.

Reply:
I think, when writing a new config, the first step would be to explicitly call the config destructor and then the constructor:

w.data().~S3Config();
::new(&w.data()) S3Config;